sycl/source/detail/kernel_program_cache.hpp

```diff
@@ -82,9 +75,9 @@ class KernelProgramCache {
     return {MKernelsPerProgramCache, MKernelsPerProgramCacheMutex};
   }
 
-  template <class Predicate> void waitUntilBuilt(Predicate Pred) const {
-    std::unique_lock<std::mutex> Lock(MBuildCVMutex);
-
+  template <class Predicate>
+  void waitUntilBuilt(std::unique_lock<std::mutex> &Lock,
```

Comment:
I think the concurrent cache management is quite complex piece of design which is worth a section in the design doc. Can you guys take an effort to write one? That would really help to do a good review.

Reply:
We can prepare a doc when @s-kanaev will be available, as he developed this solution. For now, my understanding is the following. 
When we want to get some kernel or program from cache, we first lock cache and check in which build state kernel or program is right now (Done, Failed or In Progress). 
If it is already In Progress state in the cache, when some other thread already added kernel or program to cache, in this case we move current thread to waiting state. It will be woken up by notification from other thread that build was Done or build was Failed. If build was Done, we return build result, otherwise (when build was Failed), we try one more time to build it successfully.
If there was no information about the kernel or program in the cache in In Progress state, we need to simply build it, when save the result and move build state to Done, when current thread notifies other threads that build was completed and they can wake up.
Above is explanation of how ```getOrBuild(...)``` function from ```program_manager.cpp``` works, deadlock was happened in this function. Wait and notification mechanisms utilize ```std::condition variable```, build result stores to atomic variables ```BuildResult->Ptr``` and ```BuildResult->State```.

Reply:
@kbobrovs The idea is described by @dm-vodopyanov quite correctly.